File: src/main/java/easy/TimeRequiredToBuy.java
package easy;

import java.util.ArrayList;
import java.util.List;

class Person {
  int vouchers;
  int position;
  boolean isOriginal;

  public Person(int index, int value, int k) {
    vouchers = value;
    position = index;
    isOriginal = index == k;
  }

  public void purchaseTicket() {
    if (vouchers > 0) {
      vouchers--;
    }
  }
}

public class TimeRequiredToBuy {
  public int solve(int[] tickets, int k) {
    int time = 0;
    List<Person> queue = new ArrayList<>();
    for (int i = 0; i < tickets.length; i++) {
      queue.add(new Person(i, tickets[i], k));
    }
    while (!queue.isEmpty()) {
      Person person = queue.remove(0);
      if (person.vouchers > 0) {
        person.purchaseTicket();
        time++;
      }
      if (person.isOriginal && person.vouchers == 0) break;
      if (person.vouchers > 0) queue.add(person);
    }
    return time;
  }
}

File: src/easy/time-required-to-buy.ts
class Ticket {}
class Person {
	position: number;
	tickets: Ticket[];
	isTarget: boolean;
	constructor(index: number, value: number, k: number) {
		this.isTarget = k === index;
		this.position = index;
		this.tickets = Array.from({ length: value }, () => new Ticket());
	}
	totalTickets() {
		return this.tickets.length;
	}
	canPurchaseTickets(): boolean {
		return this.totalTickets() > 0;
	}
	purchaseTicket() {
		if (this.canPurchaseTickets()) return this.tickets.shift();
	}
}
export const timeRequiredToBuy = (tickets: number[], k: number): number => {
	let time = 0;
	const queue = tickets.map((value, index) => new Person(index, value, k));

	while (true) {
		const person = queue.shift();
		if (!person || (person.isTarget && !person.canPurchaseTickets())) break;

		if (person.canPurchaseTickets()) {
			person.purchaseTicket();
			time++;
			queue.push(person);
		}
	}
	return time;
};
